roachprod: add Pyroscope continuous profiling support

herkolategan opened the pull request:
This commit introduces a full set of commands for managing [Pyroscope]￼
continuous profiling on roachprod clusters. The setup commands use Docker
Compose to deploy the Pyroscope stack required for continuous profiling.

A corresponding set of commands for adding scrape targets is also included. The
cluster (node) hosting Pyroscope does not need to be the same cluster where
Cockroach is running.

Nodes can be added or removed, triggering an automatic reload of the scrape
configuration. This enables dynamic adjustments to manage the overhead and
storage impact of continuous profiling.

[Pyroscope]: https://github.com/grafana/pyroscope

Release note: None
Epic: None

herkolategan committed:
roachprod: fix typo in registry filename

Fixed typo in registry.go filename (was resgistry.go).

Release note: None
Epic: None

herkolategan committed:
roachprod: add alloy config utility

This commit introduces the Pyroscope config package, which provides
functionality for managing Grafana Alloy configurations.

Release note: None
Epic: None

herkolategan committed:
roachprod: add `GetString` helper to `SyncedCluster`

Previously, `SyncedCluster` did not provide a convenient way to retrieve the
contents of a file in the form of a string. This change adds the `GetString`
method to `SyncedCluster` to retrieve file contents from a remote node. This
method complements the existing `PutString` method.

Release note: None
Epic: None

herkolategan committed:
roachprod: add Pyroscope continuous profiling support

This commit introduces a full set of commands for managing Pyroscope [1]￼
continuous profiling on roachprod clusters. The setup commands use Docker
Compose to deploy the Pyroscope stack required for continuous profiling.

A corresponding set of commands for adding scrape targets is also included. The
cluster (node) hosting Pyroscope does not need to be the same cluster where
Cockroach is running.

Nodes can be added or removed, triggering an automatic reload of the scrape
configuration. This enables dynamic adjustments to manage the overhead and
storage impact of continuous profiling.

[1]: https://github.com/grafana/pyroscope

Release note: None
Epic: None

golgeek reviewed (approved):
I assume if this proves useful in the scale test, we'll try to move to a central Pyroscope (accessible from our global Grafana) architecture?

herkolategan committed:
roachprod: improve alloy scrape config

Set a delta for scraping and increase the interval to 1m.

Epic: None
Release note: None

herkolategan committed:
roachprod: pyroscope set db path

In order to ensure we store the db data rather than rely on the secondary
storage blocks. This change ensures we can restart the docker stack without
losing data.

Epic: None
Release note: None

herkolategan committed:
roachprod: pyroscope limits

To prevent pyroscope from being overloaded that results in an OOM we set a query
limit of 2 hours.

The ingestion rate is also update to allow for profiling larger clusters.

Epic: None
Release note: None

herkolategan commented:
TFTR!

bors r=golgeek
